azuredevops/internal/service/servicehook/pipelines_publisher.go

```diff
@@ -0,0 +1,182 @@
+package servicehook
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+type pipelineEvent string
+
+const (
+	stageStateChanged pipelineEvent = "StageStateChanged"
+	runStateChanged   pipelineEvent = "RunStateChanged"
+)
+
+type pipelineEventType string
+
+const (
+	stageStateChangedEventType pipelineEventType = "ms.vss-pipelines.stage-state-changed-event"
+	runStateChangedEventType   pipelineEventType = "ms.vss-pipelines.run-state-changed-event"
+)
+
+var pipelineEvent2apiType = map[pipelineEvent]pipelineEventType{
+	stageStateChanged: stageStateChangedEventType,
+	runStateChanged:   runStateChangedEventType,
+}
+
+var apiType2pipelineEvent = map[pipelineEventType]pipelineEvent{
+	stageStateChangedEventType: stageStateChanged,
+	runStateChangedEventType:   runStateChanged,
+}
+
+func genPipelinesPublisherSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"published_event": {
```

Comment:
Add `AtLeastOneOf:  []string{"stage_state_changed_event, run_state_changed_event"},`  to the `stage_state_changed_event` block and we can remove the `published_event` scheam`. The publish event type can be determined via `d.GetOkExist("stage_state_changed_event")`. Also the `CustomDiff ` can be deleted too.

Reply:
I kept the `published_event` so user could set the event without any filtering like:
```hcl
resource "azuredevops_servicehook_storage_queue_pipelines" "example" {
  project_id   = azuredevops_project.example.id
  account_name = azurerm_storage_account.example.name
  account_key  = azurerm_storage_account.example.primary_access_key 
  queue_name  = azurerm_storage_queue.example.name
  visi_timeout = 30
  published_event = "RunStateChanged"
}
``` 
This way the event config would assume all default values.
What do you suggest?

the user could set an empty block like: 
```hcl
resource "azuredevops_servicehook_storage_queue_pipelines" "example" {
  project_id   = azuredevops_project.example.id
  account_name = azurerm_storage_account.example.name
  account_key  = azurerm_storage_account.example.primary_access_key 
  queue_name  = azurerm_storage_queue.example.name
  visi_timeout = 30
  run_state_changed_event {}
}
``` 

or maybe set an argument inside the block to indicate that any event will trigger the action:
```hcl
resource "azuredevops_servicehook_storage_queue_pipelines" "example" {
  project_id   = azuredevops_project.example.id
  account_name = azurerm_storage_account.example.name
  account_key  = azurerm_storage_account.example.primary_access_key 
  queue_name  = azurerm_storage_queue.example.name
  visi_timeout = 30
  run_state_changed_event {
      any = true
  }
}
```

Reply:
@xuzhang3 ok, so if I understood it correctly we can keep the resource as it is?

Reply:
I still think we can remove the `event_type`, the configuration block can represent the `event_type`. It is a bit redundant with both `event_type` and configuration block supported

Reply:
@xuzhang3 
submitted suggested changes :)